### DISCUSSION TITLE
source code of the markdown editor

### DISCUSSION THREAD

**Discussion by lucasjinreal** [emilk/egui#2595] [Q&A]
hi, where can I get the markdown editor source code?

I want using egui implement a GitNote app which take notes in markdown with git synchonization.

**Comment by coderedart** [comment]
all egui.rs examples have the source code linked at the bottom as shown in screenshot.
![image](https://user-images.githubusercontent.com/24411704/212545520-83a4c7ef-bd50-4d43-bd46-5720a32db0cc.png)

https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/easy_mark/easy_mark_editor.rs

**Comment by lucasjinreal** [comment]
@coderedart thanks, I have some further demands based on this markdown editor:

1. I need add a insert table functionality, how does it could be properly implemented?
2. I want listen to the copyboard, and read the content if it's image or not, if image, then try upload this image and get an url, does there a way to listen to it?


BTW, does the editor support Latex rendering?

**Comment by coderedart** [comment]
egui has a `Grid` widget which you can use to emulate tables. 

i don't think egui supports non-text clipboard input yet. but it supports "dragged files" and you can drag/drop an image. then, the networking part is completely up to you. egui only deals with gui. there is a small helper crate for http https://docs.rs/ehttp/latest/ehttp/

egui doesn't support any format. it just provides some basic functionality like text with different sizes / bold text / etc.. in the markdown example, it parses the markdown and then tries to create a "rendered look" using different sized texts / indents / other widgets etc...
you need to manually write similar code to parse/emulate latex rendering.

**Comment by lucasjinreal** [comment]
@coderedart thanks, does there anyway to guide me to the rendering on latex inside rust? does there any useful lib able to do that?
<img width="939" alt="image" src="https://user-images.githubusercontent.com/21303438/212663065-612c642f-7049-4aa9-a387-54eb5cb84529.png">

lt looks like image also doesn't support yet. Will it support image rendering at least for markdown?

**Reply by coderedart** [reply]
i don't really know about latex much. but you can search on crates.io . 

> lt looks like image also doesn't support yet. Will it support image rendering at least for markdown?

no. egui doesn't support markdown at all. you are looking at an "example" which showcases a combination of egui's basic widgets, to make it look like markdown rendered text. but if you wanted the images, you can probably write some egui code yourself.

although, if Latex is your main usecase, you might want to go with an existing mature framework.
